skopt/space/space.py

```diff
@@ -84,11 +90,44 @@ def check_dimension(dimension, transform=None):
     dimension : Dimension
         Dimension instance.
     """
+    old_dim = _check_dimension_old(dimension, transform=transform)
+    try:
+        new_dim = _check_dimension(dimension, transform=transform)
+    except Exception as err:
+        new_dim = f"<{err.__class__.__name__}: {err}>"
+    if new_dim != old_dim:
+        warnings.warn(f"Dimension {dimension} was inferred to {old_dim}. In "
+                      "upcoming versions of scikit-optimize, it will be "
+                      f"inferred to {new_dim}. See the documentation of the "
+                      "check_dimension function for the upcoming API.")
+    return old_dim
+
+
+def _check_dimension(dimension, transform=None):
+    if isinstance(dimension, Dimension):
+        return dimension
+    elif isinstance(dimension, (list, np.ndarray)):
```

Comment:
I don't like that a tuple of strings is no longer supported. Why does it need to be a list?

Reply:
The idea was to tighten the interface to prevent user error leading to valid, but unwanted, dimension objects. This is most surely inspired by my uses of type-checkers in python, as the above makes the signature of `check_dimension` not depend on the actual values contained in the passed sequence:
- `list` becomes a `Categorical`
- `tuple[int, int]` (and the 3- and 4-long forms) become `Integer`
- `tuple[int, float]`, `tuple[float, int]` and `tuple[float, float]` (and there 3 and 4-long forms) become `Real`

It was also to avoid an incorrect tuple silently becoming a `Categorical`, e.g.
- `(0, 1, "uniform", base)` where `base="10"` was no parsed from a data source, silently returning a `Categorical` instead of the intended `Integer`
- `(0, 1, *params)` where `params=(5, "log-uniform", 2)` because the semantic/content of `params` is wrong, silently falling back to a `Categorical`
- `[True, False]` being inferred to an `Integer` (though the old implementation has a dedicated branch to avoid that).

We could make it so that `tuple`s that fail the `Integer`/`Real` inference default to a `Categorical` as before. This would also allow to use any `Iterable`, like so:
```python
if isinstance(dimension, tuple) and 2 <= len(dimension) <= 4:
    low, high, *args = dimension
    if (not args or isinstance(args[0], str)) and (len(args) < 2 or isinstance(args[1], int)):
        if isinstance(low, numbers.Integral) and isinstance(high,  numbers.Integral):
              return Integer(...)
        elif isinstance(low, numbers.Real) and isinstance(high, numbers.Real):
            return Real(...)
# not an else, fallback
if isinstance(dimension, Iterable):
    return Categorical(...)
raise ValueError(...)
```

Do you prefer that other type-tightness ?